src/zmalloc.c

```diff
@@ -121,6 +126,34 @@ void *zmalloc(size_t size) {
     return ptr;
 }
 
+/* Try allocating memory, and return NULL if failed.
+ * '*usable' is set to the usable size if non NULL. */
+void *ztryallignedmalloc_usable(size_t alignment, size_t size, size_t *usable) {
+    ASSERT_NO_SIZE_OVERFLOW(size);
+    void *ptr;
+    int ret = posix_memalign(&ptr, alignment, size);
+    if (!ptr || ret == EINVAL || ret == ENOMEM) return NULL;
+
+#ifdef HAVE_MALLOC_SIZE
+    size = zmalloc_size(ptr);
+    update_zmalloc_stat_alloc(size);
+    if (usable) *usable = size;
+    return ptr;
+#else
+    *((size_t*)ptr) = size;
+    update_zmalloc_stat_alloc(size+PREFIX_SIZE);
+    if (usable) *usable = size;
+    return (char*)ptr+PREFIX_SIZE;
+#endif
```

Comment:
I don't see how aligned allocation can work with no `HAVE_MALLOC_SIZE`, since by definition we're breaking alignment with this pointer manipulation.

Reply:
they can work, if we modify the code to work around that (allocate a bit more extra).
i.e. add `max(sizeof(size_t), alignment)`.

note that if the user uses his own allocator, he may have to keep the size somewhere too (there's no support for malloc_size), unless he knows the allocation size implicitly somehow (they're always the same size)

Reply:
I guess alignment if often page sized, this means in practice we'd allocate `size + page_size`, just to use `sizeof(size_t)` bytes in the end of the previous page. I think this is quite a stretch...

Reply:
i didn't consider such a big alignment request, is that common? i was thinking of 16 bytes and alike.
in that's common, we must fall back to your suggestion and let the module be able to append and deduct from our used_memory.

Reply:
@DvirDukhan any input on this?

Reply:
we are dealing with 32/64B alignments.

Reply:
so when HAVE_MALLOC_SIZE is absent, when you allocate 64 aligned to 64, we'll want to allocate 72, but due to the alignment we'll end up allocating 128 instead.
if you'll ask for 1024 bytes aligned to 64, we'll want to allocate 1088 and are likely to get an entry from the 1280 allocator bin.

i think i'd rather fall back to Yossi's suggestion and let the module report changes when allocating or releasing memory using external allocators.

Reply:
bringing this PR back from hibernate:
allowing aligned allocation for systems without `HAVE_MALLOC_SIZE` will cause problems with releasing the allocation, as there might be padding before the prefix.
so, reiterating the issue, we might need API for modules to expose `update_zmalloc_stat_alloc` and `update_zmalloc_stat_free` to modules
@MeirShpilraien what is your opinion?

Reply:
How about adding a callback that will be called by Redis when it wants to know the currently used memory of the module (of memory that was allocated with an external allocator)? This way Redis can use it when calculating the currently used memory. I think this is a better approach because sometimes we might use libraries that do not allow us to override the default allocator but do allow us to get used to memory. WDYT?

Reply:
i don't think it's a good idea for redis to call a module callback every time it needs to know the memory usage.
it would have been ok if it was just in order to report it in INFO, but if redis needs to take this memory usage as if it was its own memory usage, there are tons of places that get the used memory, and they currently just read from a global variable, i don't want to replace that global variable read with a method with a loop.

Reply:
I agree with @oranagra. I also don't think it's needed - well mannered libraries will anyway let you provide your own heap functions (which will map to module API heap functions).